src/qibosoq/programs/base.py

```diff
@@ -159,60 +160,78 @@ def add_pulse_to_register(self, pulse: Pulse):
         else:
             raise NotImplementedError(f"Shape {pulse} not supported!")
 
+    def execute_drive_pulse(self, pulse: Pulse, last_pulse_registered: Dict):
+        """Register a drive pulse if needed, then trigger the respective DAC.
+
+        A pulse gets register if:
+        - it didn't happen in `initialize` (`self.pulses_registered` is False)
+        - it is not identical to the last pulse registered
+
+        """
+        if not self.pulses_registered and (
+            pulse.dac not in last_pulse_registered or pulse != last_pulse_registered[pulse.dac]
+        ):
+            self.add_pulse_to_register(pulse)
+            last_pulse_registered[pulse.dac] = pulse
+        self.pulse(ch=pulse.dac)
+
+    def execute_readout_pulse(
+        self, pulse: Pulse, muxed_pulses_executed: List[Pulse], muxed_ro_executed_indexes: List[int]
+    ):
+        """Register a readout pulse and perform a measurement."""
+        adcs = []
+        if self.is_mux:
+            if pulse not in muxed_pulses_executed:
+                for idx, mux_time in enumerate(self.multi_ro_pulses):
+                    if pulse in mux_time:
+                        idx_mux = idx
+                        break
+                self.add_muxed_readout_to_register(self.multi_ro_pulses[idx_mux])
+                muxed_ro_executed_indexes.append(idx_mux)
+                for ro_pulse in self.multi_ro_pulses[idx_mux]:
+                    adcs.append(ro_pulse.adc)
+                    muxed_pulses_executed.append(ro_pulse)
+            else:
+                return
```

Comment:
```suggestion
            if pulse in muxed_pulses_executed:
                return
            
            for idx, mux_time in enumerate(self.multi_ro_pulses):
                if pulse in mux_time:
                    idx_mux = idx
                    break
            self.add_muxed_readout_to_register(self.multi_ro_pulses[idx_mux])
            muxed_ro_executed_indexes.append(idx_mux)
            for ro_pulse in self.multi_ro_pulses[idx_mux]:
                adcs.append(ro_pulse.adc)
                muxed_pulses_executed.append(ro_pulse)
```

This is already a way to denest a bit, and it also has a name, _early return_.